crates/katana/primitives/src/genesis/mod.rs

```diff
@@ -0,0 +1,207 @@
+    pub storage: Option<HashMap<StorageKey, StorageValue>>,
+}
+
+impl GenesisAccount {
+    pub fn new(private_key: FieldElement, class_hash: ClassHash) -> (ContractAddress, Self) {
+        let public_key = public_key_from_private_key(private_key);
+        let address = get_contract_address(
+            FieldElement::from(666u32),
+            class_hash,
+            &[public_key],
+            FieldElement::ZERO,
+        );
+
+        (
+            ContractAddress::from(address),
+            Self { private_key, public_key, class_hash, ..Default::default() },
+        )
+    }
+
+    pub fn new_with_balance(
+        private_key: FieldElement,
+        class_hash: ClassHash,
+        balance: U256,
+    ) -> (ContractAddress, Self) {
+        let (address, account) = Self::new(private_key, class_hash);
+        (address, Self { balance, ..account })
+    }
+}
+
+#[serde_with::serde_as]
+#[derive(Debug, serde::Serialize)]
+pub struct FeeTokenConfig {
+    /// The name of the fee token.
+    pub name: String,
+    /// The symbol of the fee token.
+    pub symbol: String,
+    /// The address of the fee token contract.
+    pub address: ContractAddress,
+    /// The decimals of the fee token.
+    pub decimals: u8,
+    /// The total supply of the fee token.
+    pub total_supply: U256,
+    /// The class hash of the fee token contract.
+    #[serde_as(as = "UfeHex")]
+    pub class_hash: ClassHash,
+}
+
+#[serde_with::serde_as]
+#[derive(Debug, serde::Serialize, PartialEq, Eq)]
+pub struct GenesisClass {
+    /// The compiled class hash of the contract class.
+    #[serde_as(as = "UfeHex")]
+    pub compiled_class_hash: CompiledClassHash,
+    /// The casm class definition.
+    #[serde(skip_serializing)]
+    pub casm: CompiledContractClass,
+    /// The sierra class definition.
+    #[serde(skip_serializing)]
+    pub sierra: Option<FlattenedSierraClass>,
+}
+
+#[derive(Debug, serde::Serialize, PartialEq, Eq)]
+pub struct UniversalDeployerConfig {
+    pub class_hash: ClassHash,
+    pub address: ContractAddress,
+}
+
+/// Genesis block configuration.
+#[serde_with::serde_as]
+#[derive(Debug, serde::Serialize)]
+pub struct Genesis {
+    /// The genesis block parent hash.
+    #[serde_as(as = "UfeHex")]
+    pub parent_hash: BlockHash,
+    /// The genesis block state root.
+    #[serde_as(as = "UfeHex")]
+    pub state_root: FieldElement,
+    /// The genesis block number.
+    pub number: BlockNumber,
+    /// The genesis block timestamp.
+    pub timestamp: u128,
+    /// The genesis block L1 gas prices.
+    pub gas_prices: GasPrices,
+    /// The classes to declare in the genesis block.
+    pub classes: HashMap<ClassHash, GenesisClass>,
+    /// The fee token configuration.
+    pub fee_token: FeeTokenConfig,
+    /// The universal deployer (UDC) configuration.
+    pub universal_deployer: Option<UniversalDeployerConfig>,
+    /// The genesis accounts.
+    pub allocations: HashMap<ContractAddress, GenesisAccount>,
```

Comment:
If I understood the description and PR comment, a contract that is not an account can also be initialized in allocations, but does not have private key.

I don't see the `private_key` being optional currently, is it for a next PR?

Could I suggest to have this represented in the code by adjusting the nomenclature?
`GenesisContract` instead of `GenesisAccount`, and if we only have the private key being a differentiating field, I like the idea of having it optional. Or we may have a `account` key inside `GenesisContract` which is optional.

Reply:
> I don't see the private_key being optional currently,

ah good catch. i meant to make it optional.

> Could I suggest to have this represented in the code by adjusting the nomenclature?
GenesisContract instead of GenesisAccount, and if we only have the private key being a differentiating field, I like the idea of having it optional. Or we may have a account key inside GenesisContract which is optional.

in that case might be more cleaner if we just have separate types for account / non-account contracts instead of having everything optional? ofc from the point of the state there's not distinction between either kind of contracts, just for clarity in the code itself.

Reply:
> in that case might be more cleaner if we just have separate types for account / non-account contracts instead of having everything optional? ofc from the point of the state there's not distinction between either kind of contracts, just for clarity in the code itself.

Yes exactly. It would be great. As you mentioned, the state only sees contracts, and having separated types for account / non-account contract can help clarifying this separation.

First I was thinking about having a different entry in the JSON for accounts specifically. Which can help one separating the concerns of state related allocations and private key associated to an account. But this may be confusing, what's your take on that?

Reply:
> First I was thinking about having a different entry in the JSON for accounts specifically. Which can help one separating the concerns of state related allocations and private key associated to an account. But this may be confusing, what's your take on that?

Do you mean by having two separate 'allocations' field in the JSON for accounts-specific and generic contracts parameters?

Yeah, I initially was thinking about that approach as well but i can kinda see how user might get confused by the separation, though not very certain if it would really be perceived as 'confusing'. 

I was not keen on that approach at first because seemed somewhat redundant but it might be good to make it distinct.

Maybe something like :

```jsonc
{
  // for account allocations
  "accounts": {
    "0x1": { .. }
  },
  // for generic contracts
  "contracts": {
    "0x2": { .. }
  }
}
```

where both would share similar fields like `storage`, `nonce` but account would have extra fields (ie `privateKey`). wdyt ?

Reply:
> > First I was thinking about having a different entry in the JSON for accounts specifically. Which can help one separating the concerns of state related allocations and private key associated to an account. But this may be confusing, what's your take on that?
> 
> Do you mean by having two separate 'allocations' field in the JSON for accounts-specific and generic contracts parameters?
> 
> Yeah, I initially was thinking about that approach as well but i can kinda see how user might get confused by the separation, though not very certain if it would really be perceived as 'confusing'.
> 
> I was not keen on that approach at first because seemed somewhat redundant but it might be good to make it distinct.
> 
> Maybe something like :
> 
> ```js
> {
>   // for account allocations
>   "accounts": {
>     "0x1": { .. }
>   },
>   // for generic contracts
>   "contracts": {
>     "0x2": { .. }
>   }
> }
> ```
> 
> where both would share similar fields like `storage`, `nonce` but account would have extra fields (ie `privateKey`). wdyt ?

This yes, or we can have allocations that are for state stuff (`storage`, `nonce` etc..) without distinction between accounts and regular contracts.

And then we may have a dedicated JSON key for `accounts`:
```json
  "allocations": {
	  "0x44": {
		  "storage": {
			  "0x1": "0x1",
			  "0x2": "0x2"
		  }
	  },
	  "0x88": {
		  "nonce": "0x1",
		  "storage": {
			  "0x1": "0x11",
			  "0x2": "0x22"
		  }
	  },
  },
  "accounts": {
          "0x44": {
		  "privateKey": "0x1",
          },
  }
```
Doing so, we have a repetition of the address, but we can make both evolve quite independently.

But if you prefer to keep it DRY, your approach with `accounts` and `contracts` makes sense to me too.

Reply:
> ```json
>   "allocations": {
> 	  "0x44": {
> 		  "storage": {
> 			  "0x1": "0x1",
> 			  "0x2": "0x2"
> 		  }
> 	  },
> 	  "0x88": {
> 		  "nonce": "0x1",
> 		  "storage": {
> 			  "0x1": "0x11",
> 			  "0x2": "0x22"
> 		  }
> 	  },
>   },
>   "accounts": {
>           "0x44": {
> 		  "privateKey": "0x1",
>           },
>   }
> ```
> 
> Doing so, we have a repetition of the address, but we can make both evolve quite independently.

I think this approach seems nice, the repetition isn't exactly a big issue. Let's just go with this approach for initial implementation. We can figure out the better format as we go and don't have to be too particular for now (which I'm usually guilty of).

Reply:
> > ```json
> >   "allocations": {
> > 	  "0x44": {
> > 		  "storage": {
> > 			  "0x1": "0x1",
> > 			  "0x2": "0x2"
> > 		  }
> > 	  },
> > 	  "0x88": {
> > 		  "nonce": "0x1",
> > 		  "storage": {
> > 			  "0x1": "0x11",
> > 			  "0x2": "0x22"
> > 		  }
> > 	  },
> >   },
> >   "accounts": {
> >           "0x44": {
> > 		  "privateKey": "0x1",
> >           },
> >   }
> > ```

nvm just realized this separation makes reading the config a bit of a hassle as you have to look at the at both fields back and forth instead of just having it in the same object.